src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/MarkupTransitionCompletionItemProvider.cs

```diff
@@ -0,0 +1,99 @@
+        public static RazorCompletionItem MarkupTransitionCompletionItem
+        {
+            get
+            {
+                if (_markupTransitionCompletionItem == null)
+                {
+                    var completionDisplayText = SyntaxConstants.TextTagName;
+                    _markupTransitionCompletionItem = new RazorCompletionItem(
+                        completionDisplayText,
+                        completionDisplayText,
+                        RazorCompletionItemKind.MarkupTransition,
+                        ElementCommitCharacters);
+                    var completionDescription = new MarkupTransitionCompletionDescription(CSharpCodeParser.AddMarkupTransitionDescriptor);
+                    _markupTransitionCompletionItem.SetMarkupTransitionCompletionDescription(completionDescription);
+                }
+
+                return _markupTransitionCompletionItem;
+            }
+        }
+
+        public MarkupTransitionCompletionItemProvider(HtmlFactsService htmlFactsService)
+        {
+            if (htmlFactsService is null)
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorSyntaxTree syntaxTree, TagHelperDocumentContext tagHelperDocumentContext, SourceSpan location)
+        {
+            if (syntaxTree is null)
+            {
+                throw new ArgumentNullException(nameof(syntaxTree));
+            }
+
+            var change = new SourceChange(location, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            if (!AtMarkupTransitionCompletionPoint(owner))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var parent = owner.Parent;
+            if (!_htmlFactsService.TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) ||
+                !containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+            {
+                return Array.Empty<RazorCompletionItem>();
+            }
+
+            var completions = new List<RazorCompletionItem>() { MarkupTransitionCompletionItem };
+            return completions;
+        }
+
+        // Internal for testing
+        internal static bool AtMarkupTransitionCompletionPoint(RazorSyntaxNode owner)
+        {
+            // Only provide IntelliSense for C# code blocks, of the form:
+            // @{ }, @code{ }, @functions{ }, @if(true){ }
+            var closestSignificantAncestor = owner.Ancestors().FirstOrDefault(node =>
+            {
+                return (node is MarkupElementSyntax markupNode && markupNode.ChildNodes().Count != 1) ||
```

Comment:
This may be terser but is a painnnn to debug. I'd break these into separate if statements:

```C#
if (node is MarkupElementSyntax markupNode && markupNode.ChildNodes().Count != 1)
{
    return true
}

if (node is MarkupMinimizedAttributeBlockSyntax)
{
    return true;
}

if (node is CSharpCodeBlockSyntax)
{
    return true;
}

return false;
```